### ISSUE TITLE
Language Feast: Haskell

### ISSUE THREAD

**Issue by Oleksii-Kshenskyi** [Oleksii-Kshenskyi/btagsil#21]
An attempt to start up a BTAGSIL implementation in Haskell.
This implementation should have the following:
- You start with bare fists as your weapon.
- Four locations: Forest, Square, Weapon Shop, Cave. You can get a new weapon in the Weapon Shop and fight a monster in the Cave.
- Three weapons: Axe, Sword, Bow: all with different damage stats.
- Navigational commands: you can ask where you are, where you can go, move between neighboring locations, etc.
- On the square, you can talk to the guard who can tell you a random fact or joke.

Monitor your experience and impressions from developing this kind of application in Haskell, and once the implementation is finished, think about whether you liked it enough to start a gigantic application of this style in it or not.

**Comment by Oleksii-Kshenskyi** [comment]
The Haskell experiment is over. Unfortunately, it was unsuccessful.
1. I have only been able to implement the most basic possible conditional parser, the concept of an action and three actions: exit, echo and empty. No game world was actually implemented.
2. The reason for that is that having an all-encompassing World object in Haskell is a nightmare as everything is immutable and pure. Therefore if you need an immense object where you need to be able to change a single part of it, you need to know one of (or preferably all of) the complicated concepts that allow for that: State Monad, Functional Reactive Programming, Lenses are the three that I've been able to find. Even just grasping any of these is several weeks of work.
3. The parser itself is extremely basic and doesn't do much of anything. If I were to create a proper parser that is flexible and easily modifiable, I would need to grasp Parsec, Monad, Applicative, several functional operators that I don't understand, etc.
4. At this point I'm pretty sure that creating a proper game in Haskell would require advanced knowledge of the language and pure functional concepts that would take me months to fully grasp. I would be open to that kind of long experiment, but only as a side project.
5. For BTAGSIL, I'm looking for a solid harmony between knowing what I'm doing and learning a bunch of new things, Haskell is obviously not the choice here.

**Comment by Oleksii-Kshenskyi** [comment]
Re-opening the Haskell prototype to try one more time. The logic is as follows: as for now, the main and obvious leader in terms of my positive impressions of all the prototypes currently is Scala. Scala is a great option because it's statically typed, it's as OOP and/or functional as you want it to be, it can either be Java or Haskell, the choice is yours, while Haskell is purely functional and that's it. However, a couple inefficiencies in the Scala Language Feast (#24) were still annoying. Therefore, I'd like to try one final time and implement the Haskell prototype to see if I can overcome the barrier to entry. If yes, Haskell is going to be an amazing choice. If the prototype goes bust a second time, it's going to be Scala.

**Comment by Oleksii-Kshenskyi** [comment]
This damn Language Feast issue is taking me ages to complete. The reasons being:

Haskell seems exceptional by such metrics as: code clarity + cleanliness, being a compiled and statically typed language, being quite fast and quite well supported by the community. However, the things that some people have claimed made Haskell a "toy language", such as: the contagion of the IO monad (the most important one), forcing developers to try and contain that contagion, needing a separate gigantic library (lenses) for proper getters and setters to complex structures, making the issue even worse if said complex structure needs mutability, etc.

I'll try and intensify the efforts to complete the prototype in the next couple of days, but the probability of choosing Haskell as the implementation language for BTAGSIL doesn't look too good, as it's just too annoying to deal with in many regards.

**Comment by Oleksii-Kshenskyi** [comment]
The monster/cave related functionality is unfinished, but as the [Language Feast](https://github.com/Oleksii-Kshenskyi/btagsil/milestone/3) decision (meaning the decision which language/technology to use for BTAGSIL) has already been made, the issue can be closed. The only thing that's left is to document the actual result of the milestone is a separate issue.
